¡Hola, Angel!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un enfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">

<div class="alert alert-block alert-success">
<b>Review General iteración 1) </b> <a class="tocSkip"></a>
    
Angel, has hecho un excelente trabajo con el desarrollo del proyecto, cada vez más cercas de convertirte en un cientifico de datos. Realizaste la carga de bases, su análisis inicial,el desarrollo de los modelos, el calculo del RMSE además lo replicate para todas la regiones y concluiste con el análisis de ganancias. Solamente te dejo algunos comentarios para complementar el análisis.

Sigue con el excelente trabajo!

# 1. Preparación de los datos

In [1]:
#Importar librerías

import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [2]:
#Cargar los datos 

data_0 = pd.read_csv('/datasets/geo_data_0.csv')
data_1 = pd.read_csv('/datasets/geo_data_1.csv')
data_2 = pd.read_csv('/datasets/geo_data_2.csv')

In [3]:
#Eliminar ID y revisar info básica

for i, df in enumerate([data_0, data_1, data_2]):
    df.drop('id', axis=1, inplace=True)
    print(f"Región {i}: {df.shape}")

Región 0: (100000, 4)
Región 1: (100000, 4)
Región 2: (100000, 4)


In [4]:
data_0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float64
 1   f1       100000 non-null  float64
 2   f2       100000 non-null  float64
 3   product  100000 non-null  float64
dtypes: float64(4)
memory usage: 3.1 MB


In [5]:
data_0.head()

,f0,f1,f2,product
0,0.705745,-0.497823,1.221170,105.280062
1,1.334711,-0.340164,4.365080,73.037750
2,1.022732,0.151990,1.419926,85.265647
3,-0.032172,0.139033,2.978566,168.620776
4,1.988431,0.155413,4.751769,154.036647


In [6]:
data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float64
 1   f1       100000 non-null  float64
 2   f2       100000 non-null  float64
 3   product  100000 non-null  float64
dtypes: float64(4)
memory usage: 3.1 MB


In [7]:
data_1.head()

,f0,f1,f2,product
0,-15.001348,-8.276000,-0.005876,3.179103
1,14.272088,-3.475083,0.999183,26.953261
2,6.263187,-5.948386,5.001160,134.766305
3,-13.081196,-11.506057,4.999415,137.945408
4,12.702195,-8.147433,5.004363,134.766305


In [8]:
data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   f0       100000 non-null  float64
 1   f1       100000 non-null  float64
 2   f2       100000 non-null  float64
 3   product  100000 non-null  float64
dtypes: float64(4)
memory usage: 3.1 MB


In [9]:
data_2.head()

,f0,f1,f2,product
0,-1.146987,0.963328,-0.828965,27.758673
1,0.262778,0.269839,-2.530187,56.069697
2,0.194587,0.289035,-5.586433,62.871910
3,2.236060,-0.553760,0.930038,114.572842
4,-0.515993,1.716266,5.899011,149.600746


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Excelente trabajo con la carga de la información y su exploración inicial.  

Solamente te recomiendo complementar con un análisis de los registros duplicados que como aprendiste en cursos pasados sirve para asegurar que nuestros resultados no estarán sesgados

# 2. Entrenamiento y prueba del modelo para cada región

In [10]:
def train_and_evaluate(df):
    features = df.drop('product', axis=1)
    target = df['product']
    
    # Dividir los datos 
    features_train, features_valid, target_train, target_valid = train_test_split(
        features, target, test_size=0.25, random_state=12345)
    
    # Entrenar modelo
    model = LinearRegression()
    model.fit(features_train, target_train)
    
    # Predicciones
    predictions = model.predict(features_valid)
    predictions = pd.Series(predictions, index=target_valid.index)

    #Calculo del error
    rmse = mean_squared_error(target_valid, predictions) ** 0.5
    avg_predicted = predictions.mean()
    
    return target_valid, predictions, rmse, avg_predicted

# Ejecutar para las 3 regiones

results = {}
for i, df in enumerate([data_0, data_1, data_2]):
    target_valid, pred, rmse, avg = train_and_evaluate(df)
    results[i] = {'target': target_valid, 'predictions': pred}
    print(f"Región {i}: RMSE = {rmse:.2f}, Promedio predicho = {avg:.2f}")

Región 0: RMSE = 37.58, Promedio predicho = 92.59
Región 1: RMSE = 0.89, Promedio predicho = 68.73
Región 2: RMSE = 40.03, Promedio predicho = 94.97


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Angel, implementaste el modelo de regresión lineal de forma excelente para los 3 conjuntos de datos! En este punto ya puedes obervar cuál es el modelo con el mayor $R^2$ y menor RMSE. 
    

# 3. Cálculo de ganancias

In [11]:
#Definir variables 

presupuesto = 100_000_000
pozos_a_abrir = 200
ingresos_por_unidad = 4500

# Cálculo del volumen mínimo necesario por pozo
costo_por_pozo= presupuesto / pozos_a_abrir
unidades_necesarias = costo_por_pozo / ingresos_por_unidad

print(f"Volumen mínimo necesario por pozo: {unidades_necesarias:.2f} mil barriles")

Volumen mínimo necesario por pozo: 111.11 mil barriles


Al comparar este valor con el promedio de reservas de las regiones (Región 0: 92.5, Región 1: 68.7, Región 2: 95), se observa que ninguna región alcanza el punto de equilibrio de forma natural.

# 4. Cálculo de ganancia 

In [12]:
#Definición de la funcion para el cálculo

def calculate_profit(target, predictions, count):
    
    # Seleccionar los índices de las predicciones más altas
    preds_sorted = predictions.sort_values(ascending=False)
    selected_targets = target[preds_sorted.index][:count]
    
    total_volume = selected_targets.sum()
    revenue = total_volume * ingresos_por_unidad
    return revenue - presupuesto


In [13]:
#Predicciones para los pozos para las 3 regiones

for i in range(3):
    target = results[i]['target']
    predictions = results[i]['predictions']
    
    # Cálculola ganancia
    ganancia = calculate_profit(target, predictions, pozos_a_abrir)
    top_preds = predictions.sort_values(ascending=False)[:pozos_a_abrir]
    volumen_real_top = target[top_preds.index].sum()
    
    print(f"--- REGIÓN {i} ---")
    print(f"Volumen total real de los 200 mejores pozos: {volumen_real_top:.2f} mil barriles")
    print(f"Ganancia potencial: ${ganancia:,.2f} USD\n")

--- REGIÓN 0 ---
Volumen total real de los 200 mejores pozos: 29601.84 mil barriles
Ganancia potencial: $33,208,260.43 USD

--- REGIÓN 1 ---
Volumen total real de los 200 mejores pozos: 27589.08 mil barriles
Ganancia potencial: $24,150,866.97 USD

--- REGIÓN 2 ---
Volumen total real de los 200 mejores pozos: 28245.22 mil barriles
Ganancia potencial: $27,103,499.64 USD



Según el beneficio máximo obtenido en el cálculo, la región recomendada para el desarrollo es la región 0, debido a que presenta el mayor margen de beneficio. Aunque esta región destaca por su rentabilidad potencial, sin embargo, dicha región presenta un RMSE elevado, lo que implica que la incertidumbre en las prediccion es significativa.

# 5. Cálculo de riesgo y ganancia para cada región

In [14]:
import numpy as np

state = np.random.RandomState(12345)

for i in range(3):
    target = results[i]['target']
    predictions = results[i]['predictions']
    values = []
    
    for _ in range(1000):
        # Seleccionamos 500 puntos al azar (estudio geológico)
        target_subsample = target.sample(n=500, replace=True, random_state=state)
        probs_subsample = predictions[target_subsample.index]
        
        # Calculamos ganancia de los mejores 200 de esos 500
        profit = calculate_profit(target_subsample, probs_subsample, pozos_a_abrir)
        values.append(profit)
    
    values = pd.Series(values)
    mean_profit = values.mean()
    risk = (values < 0).mean() * 100
    lower = values.quantile(0.025)
    upper = values.quantile(0.975)
    
    print(f"--- REGIÓN {i} ---")
    print(f"Beneficio promedio: {mean_profit:,.2f} USD")
    print(f"Intervalo de confianza (95%): ({lower:,.2f}, {upper:,.2f})")
    print(f"Riesgo de pérdida: {risk:.2f}%")

--- REGIÓN 0 ---
Beneficio promedio: 4,259,385.27 USD
Intervalo de confianza (95%): (-1,020,900.95, 9,479,763.53)
Riesgo de pérdida: 6.00%
--- REGIÓN 1 ---
Beneficio promedio: 5,182,594.94 USD
Intervalo de confianza (95%): (1,281,232.31, 9,536,129.82)
Riesgo de pérdida: 0.30%
--- REGIÓN 2 ---
Beneficio promedio: 4,201,940.05 USD
Intervalo de confianza (95%): (-1,158,526.09, 9,896,299.40)
Riesgo de pérdida: 6.20%


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo con el calculo de las ganacias para cada una de las regiones. Posteriormente este resultados lo complementas con el análisis de riesgos


# Conclusión

Tras realizar el análisis técnico y financiero, la región recomendada para el desarrollo de los 200 pozos es la región 1. Aunque otras regiones mostraron volúmenes de reserva máximos más altos en el análisis, la región 1 es la única que cumple con los criterios de seguridad de OilyGiant ya que presenta un riesgo de 0.30%, el cual es inferior al límite máximo permitido del 2.5%. Ademas esta región obtuvo el RMSE más bajo (0.89), por ende las predicciones son precisas. Así mismo el beneficio promedio estimado mediante bootstrapping es de 5,182,594.94 USD. Dada esta informacion la opcion seleccionada en el punto 4 (región 0) se descarta.